File: src/tension_accumulator.py
"""Tension Accumulator — updates belief confidence/tension based on evidence."""

from uuid import UUID
from src.models import Evidence, Belief
from src.belief_graph import reinforce_belief, add_tension, log_contradiction, get_beliefs_above_threshold
from src.config import TENSION_INCREMENT, REVISION_THRESHOLD


async def accumulate(
    evidence_list: list[Evidence],
    interaction_id: UUID | None = None,
) -> list[Belief]:
    """Process evidence and update beliefs. Returns beliefs that crossed revision threshold."""
    triggered = []

    for ev in evidence_list:
        if not ev.belief_id or ev.stance == "neutral":
            continue

        if ev.stance == "reinforcing":
            await reinforce_belief(ev.belief_id)

        elif ev.stance == "contradicting":
            # Tension increment scaled by evidence strength — NO diminishing returns
            delta = TENSION_INCREMENT * ev.strength
            updated = await add_tension(ev.belief_id, delta)

            # Log the contradiction
            await log_contradiction(
                belief_id=ev.belief_id,
                interaction_id=interaction_id,
                evidence=ev.claim,
                tension_delta=delta,
            )

            # Check if this pushed the belief over the threshold
            if updated.tension >= REVISION_THRESHOLD:
                triggered.append(updated)

    return triggered

File: ts/src/tension-accumulator.ts
/**
 * Tension Accumulator — updates belief confidence/tension based on evidence.
 */

import {
  reinforceBelief,
  addTension,
  logContradiction,
} from "./belief-graph.js";
import { TENSION_INCREMENT, REVISION_THRESHOLD } from "./config.js";
import type { Belief, Evidence } from "./types.js";

export async function accumulate(
  evidenceList: Evidence[],
  interactionId: string | null = null,
): Promise<Belief[]> {
  const triggered: Belief[] = [];

  for (const ev of evidenceList) {
    if (!ev.belief_id || ev.stance === "neutral") continue;

    if (ev.stance === "reinforcing") {
      await reinforceBelief(ev.belief_id);
    } else if (ev.stance === "contradicting") {
      const delta = TENSION_INCREMENT * ev.strength;
      const updated = await addTension(ev.belief_id, delta);

      await logContradiction(ev.belief_id, interactionId, ev.claim, delta);

      if (updated && updated.tension >= REVISION_THRESHOLD) {
        triggered.push(updated);
      }
    }
  }

  return triggered;
}
